crates/kumod/src/smtp_server.rs

```diff
@@ -650,13 +682,37 @@ impl EsmtpListenerParams {
                         socket.set_nodelay(true)?;
                         let my_address = socket.local_addr()?;
                         let params = self.clone();
+                        let proxy_mode = proxy_mode;
                         SMTPSRV.spawn(
-                            format!("SmtpServerSession {peer_address:?}"),
+                            format!("SmtpServerSession {orig_peer_address:?}"),
                             async move {
+                                let (peer_address, client_address) = if proxy_mode.is_proxy_protocol_v2() {
+                                    match tokio::time::timeout(
+                                        Duration::from_secs(2),
+                                        extract_proxy_protocol_peer_address(&mut socket, orig_peer_address)
+                                    ).await {
+                                        Ok(Ok(addr)) => (orig_peer_address, addr),
+                                        Ok(Err(_)) | Err(_) => {
+                                            tracing::warn!("Proxy protocol v2 header invalid or timeout");
+                                            let response = format!("421 4.7.1 {} proxy protocol required\r\n", orig_peer_address);
```

Comment:
SMTP requires that the hostname be included in a 421 response

```suggestion
                                            let response = format!("421 4.7.1 {hostname} {} proxy protocol required\r\n", orig_peer_address);
```